### PR TITLE
fix: configure discard policy for WorkQueue/Interest 

### PR DESCRIPTION
**Pull request by kohlisid**

### CODE REVIEW

**Comment by vigith** [comment]
should we add a check so that nobody will turn on `DiscardOld` with `WorkQueue` *or* `DiscardNew` with `Limits` ?

**Comment by kohlisid** [comment]
In that case do you want to tie both the properties together instead of configurable?
1) If Limits/Interest is given -> use discardOld
2) If workqueue is given -> use discardNew

File: pkg/isbsvc/jetstream_service.go [hunk 1]
```diff
@@ -120,11 +120,17 @@ func (jss *jetStreamSvc) CreateBuffersAndBuckets(ctx context.Context, buffers, b
 			if !errors.Is(err, nats.ErrStreamNotFound) {
 				return fmt.Errorf("failed to query information of stream %q during buffer creating, %w", streamName, err)
 			}
+			retention := nats.RetentionPolicy(v.GetInt("stream.retention"))
+			discard := nats.DiscardPolicy(v.GetInt("stream.discardPolicy"))
+			// we cannot use DiscardNew with Limits policy.
+			if retention == nats.LimitsPolicy && discard == nats.DiscardNew {
```
Comment (Krithika3) [inline pkg/isbsvc/jetstream_service.go:126]: Shoudl we also check for the other? WorkQueue with DiscardOld?

Comment (Krithika3) [reply pkg/isbsvc/jetstream_service.go:126]: nevermind, just saw that @vigith has already made that comment.

**Comment by vigith** [comment]
> In that case do you want to tie both the properties together instead of configurable?
> 
> 1. If Limits/Interest is given -> use discardOld
> 2. If workqueue is given -> use discardNew

I would think so, @whynowy ?

File: config/install.yaml [hunk 2]
```diff
@@ -18329,6 +18329,8 @@ data:
             storage: 0
             replicas: 3
             duplicates: 60s
+            # 0: DiscardOld, 1: DiscardNew
+            discardPolicy: 0
```
Comment (Krithika3) [inline config/install.yaml:18333]: just curious, why do we have multiple configs where we have this field set?

Comment (kohlisid) [reply config/install.yaml]: We use the different installation strategies for the numaflow controller, so replicating it for all.

File: pkg/isbsvc/jetstream_service.go [hunk 3]
```diff
@@ -120,11 +120,23 @@ func (jss *jetStreamSvc) CreateBuffersAndBuckets(ctx context.Context, buffers, b
...
+			retention := nats.RetentionPolicy(v.GetInt("stream.retention"))
+			// default discard policy is set to DiscardOld
+			discard := nats.DiscardOld
+
+			// Based on the retention policy we use the following discard policy
+			// 1) Limits/Interest Policy -> DiscardOld
+			// 2) WorkQueuePolicy -> DiscardNew
+			if retention == nats.WorkQueuePolicy {
+				discard = nats.DiscardNew
+			}
```
Comment (vigith) [inline pkg/isbsvc/jetstream_service.go:133]: what about Interest? Why should it be `DiscardOld`? please document

Comment (kohlisid) [reply pkg/isbsvc/jetstream_service.go]: I went ahead as it was default, but I realised it didn't seem to be correct.
We should use it with DiscardNew instead
I have added few lines to document, please take a look @vigith

**Review by vigith — APPROVED** [review]

**Review by whynowy — APPROVED** [review]

**Pull request merged**